File: packages/schemas/ts/src/project.ts
import { z } from "zod";
import { SchemaVersionV1 } from "./schema-version";

export const ResearchPurpose = z.enum([
  "learning",
  "teaching",
  "writing",
  "research",
  "industry",
  "product",
  "strategy",
  "custom",
]);

export const ResearchPlanStatus = z.enum([
  "draft",
  "pending-approval",
  "approved",
  "rejected",
  "superseded",
]);

export const ResearchTaskType = z.enum([
  "search",
  "source-evaluation",
  "extraction",
  "entity",
  "relation",
  "validation",
  "writer",
  "synthesis",
]);

export const ResearchTaskStatus = z.enum([
  "PENDING",
  "PLANNING",
  "RUNNING",
  "VALIDATING",
  "NEEDS_REVIEW",
  "COMPLETED",
  "FAILED",
  "PAUSED",
  "CANCELLED",
]);

export const TaskDependencyCondition = z.enum(["completed", "completed-or-skipped"]);

export const ResearchRunStatus = z.enum(["running", "paused", "completed", "failed", "cancelled"]);

export const ProjectSchema = z.object({
  schema_version: SchemaVersionV1,
  project_id: z.string(),
  name: z.string().min(1),
  description: z.string().optional(),
  vault_path: z.string().optional(),
  archived: z.boolean().optional(),
  created_at: z.string(),
  updated_at: z.string(),
});
export type Project = z.infer<typeof ProjectSchema>;

export const TimeRangeSchema = z
  .object({
    from: z.string().optional(),
    to: z.string().optional(),
  })
  .strict();

export const ResearchConfigSchema = z.object({
  schema_version: SchemaVersionV1,
  config_id: z.string(),
  project_id: z.string(),
  domain: z.string().min(1),
  topic: z.string().min(1),
  purpose: ResearchPurpose,
  audience: z.string().optional(),
  depth: z.number().int().min(1).max(5),
  dimensions: z.array(z.string().min(1)).min(1),
  time_range: TimeRangeSchema.optional(),
  geographic_scope: z.string().optional(),
  languages: z.array(z.string().min(2)).min(1),
  source_types: z.array(z.string().min(1)).min(1),
  source_domains: z.array(z.string()).optional(),
  update_frequency: z.string().optional(),
  created_at: z.string().optional(),
  updated_at: z.string().optional(),
});
export type ResearchConfig = z.infer<typeof ResearchConfigSchema>;

export const PlanGeneratedBySchema = z
  .object({
    prompt_id: z.string(),
    prompt_version: z.string(),
  })
  .strict();

export const ResearchPlanSchema = z.object({
  schema_version: SchemaVersionV1,
  plan_id: z.string(),
  project_id: z.string(),
  config_id: z.string(),
  title: z.string().optional(),
  status: ResearchPlanStatus,
  section_ids: z.array(z.string()).min(1),
  generated_by: PlanGeneratedBySchema.optional(),
  approved_at: z.string().optional(),
  rejected_at: z.string().optional(),
  superseded_by: z.string().optional(),
  created_at: z.string(),
  updated_at: z.string().optional(),
});
export type ResearchPlan = z.infer<typeof ResearchPlanSchema>;

export const ResearchSectionSchema = z.object({
  schema_version: SchemaVersionV1,
  section_id: z.string(),
  plan_id: z.string(),
  title: z.string().min(1),
  dimension: z.string().optional(),
  rationale: z.string().optional(),
  order: z.number().int().min(0),
  task_ids: z.array(z.string()).optional(),
  created_at: z.string().optional(),
  updated_at: z.string().optional(),
});
export type ResearchSection = z.infer<typeof ResearchSectionSchema>;

export const ResearchTaskSchema = z.object({
  schema_version: SchemaVersionV1,
  task_id: z.string(),
  plan_id: z.string(),
  run_id: z.string().optional(),
  section_id: z.string().optional(),
  project_id: z.string(),
  type: ResearchTaskType,
  status: ResearchTaskStatus,
  idempotency_key: z.string().min(1),
  checkpoint: z.record(z.unknown()).optional(),
  retry_count: z.number().int().min(0).optional(),
  max_retries: z.number().int().min(0).optional(),
  cache_refs: z.array(z.string()).optional(),
  result_ref: z.string().optional(),
  error_ref: z.string().optional(),
  created_at: z.string(),
  updated_at: z.string(),
});
export type ResearchTask = z.infer<typeof ResearchTaskSchema>;

export const TaskDependencySchema = z.object({
  schema_version: SchemaVersionV1,
  task_id: z.string(),
  depends_on_task_id: z.string(),
  condition: TaskDependencyCondition,
  created_at: z.string().optional(),
});
export type TaskDependency = z.infer<typeof TaskDependencySchema>;

export const ResearchRunSchema = z.object({
  schema_version: SchemaVersionV1,
  run_id: z.string(),
  project_id: z.string(),
  plan_id: z.string(),
  status: ResearchRunStatus,
  config_snapshot: z.record(z.unknown()).optional(),
  plan_snapshot_ref: z.string().optional(),
  stats: z.record(z.unknown()).optional(),
  started_at: z.string(),
  finished_at: z.string().optional(),
});
export type ResearchRun = z.infer<typeof ResearchRunSchema>;

File: packages/schemas/rust/src/project.rs
//! Project, research configuration, plan, section, task, dependency, and run bindings.
//!
//! Domain records are open: unknown additive fields from newer minors are ignored,
//! matching the canonical JSON Schemas. Enums use `rename_all` to match the
//! canonical casing exactly.

use serde::{Deserialize, Serialize};

use crate::schema_version::SchemaVersionV1;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ResearchPurpose {
    Learning,
    Teaching,
    Writing,
    Research,
    Industry,
    Product,
    Strategy,
    Custom,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ResearchPlanStatus {
    Draft,
    PendingApproval,
    Approved,
    Rejected,
    Superseded,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ResearchTaskType {
    Search,
    SourceEvaluation,
    Extraction,
    Entity,
    Relation,
    Validation,
    Writer,
    Synthesis,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ResearchTaskStatus {
    Pending,
    Planning,
    Running,
    Validating,
    NeedsReview,
    Completed,
    Failed,
    Paused,
    Cancelled,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum TaskDependencyCondition {
    Completed,
    CompletedOrSkipped,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ResearchRunStatus {
    Running,
    Paused,
    Completed,
    Failed,
    Cancelled,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Project {
    pub schema_version: SchemaVersionV1,
    pub project_id: String,
    pub name: String,
    pub description: Option<String>,
    pub vault_path: Option<String>,
    pub archived: Option<bool>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TimeRange {
    pub from: Option<String>,
    pub to: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResearchConfig {
    pub schema_version: SchemaVersionV1,
    pub config_id: String,
    pub project_id: String,
    pub domain: String,
    pub topic: String,
    pub purpose: ResearchPurpose,
    pub audience: Option<String>,
    pub depth: i32,
    pub dimensions: Vec<String>,
    pub time_range: Option<TimeRange>,
    pub geographic_scope: Option<String>,
    pub languages: Vec<String>,
    pub source_types: Vec<String>,
    pub source_domains: Option<Vec<String>>,
    pub update_frequency: Option<String>,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlanGeneratedBy {
    pub prompt_id: String,
    pub prompt_version: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResearchPlan {
    pub schema_version: SchemaVersionV1,
    pub plan_id: String,
    pub project_id: String,
    pub config_id: String,
    pub title: Option<String>,
    pub status: ResearchPlanStatus,
    pub section_ids: Vec<String>,
    pub generated_by: Option<PlanGeneratedBy>,
    pub approved_at: Option<String>,
    pub rejected_at: Option<String>,
    pub superseded_by: Option<String>,
    pub created_at: String,
    pub updated_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResearchSection {
    pub schema_version: SchemaVersionV1,
    pub section_id: String,
    pub plan_id: String,
    pub title: String,
    pub dimension: Option<String>,
    pub rationale: Option<String>,
    pub order: i64,
    pub task_ids: Option<Vec<String>>,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResearchTask {
    pub schema_version: SchemaVersionV1,
    pub task_id: String,
    pub plan_id: String,
    pub run_id: Option<String>,
    pub section_id: Option<String>,
    pub project_id: String,
    #[serde(rename = "type")]
    pub task_type: ResearchTaskType,
    pub status: ResearchTaskStatus,
    pub idempotency_key: String,
    pub checkpoint: Option<serde_json::Value>,
    pub retry_count: Option<i64>,
    pub max_retries: Option<i64>,
    pub cache_refs: Option<Vec<String>>,
    pub result_ref: Option<String>,
    pub error_ref: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskDependency {
    pub schema_version: SchemaVersionV1,
    pub task_id: String,
    pub depends_on_task_id: String,
    pub condition: TaskDependencyCondition,
    pub created_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResearchRun {
    pub schema_version: SchemaVersionV1,
    pub run_id: String,
    pub project_id: String,
    pub plan_id: String,
    pub status: ResearchRunStatus,
    pub config_snapshot: Option<serde_json::Value>,
    pub plan_snapshot_ref: Option<String>,
    pub stats: Option<serde_json::Value>,
    pub started_at: String,
    pub finished_at: Option<String>,
}

use crate::ContractValid;

impl ContractValid for Project {
    fn check_contract(&self) -> Result<(), String> {
        if self.name.is_empty() {
            return Err("name must not be empty".into());
        }
        Ok(())
    }
}

impl ContractValid for ResearchConfig {
    fn check_contract(&self) -> Result<(), String> {
        if self.domain.is_empty() {
            return Err("domain must not be empty".into());
        }
        if self.topic.is_empty() {
            return Err("topic must not be empty".into());
        }
        if !(1..=5).contains(&self.depth) {
            return Err(format!("depth must be between 1 and 5, got {}", self.depth));
        }
        if self.dimensions.is_empty() || self.dimensions.iter().any(|d| d.is_empty()) {
            return Err("dimensions must have at least one non-empty item".into());
        }
        if self.languages.is_empty() || self.languages.iter().any(|l| l.len() < 2) {
            return Err("languages must have at least one item of length >= 2".into());
        }
        if self.source_types.is_empty() || self.source_types.iter().any(|s| s.is_empty()) {
            return Err("source_types must have at least one non-empty item".into());
        }
        Ok(())
    }
}

impl ContractValid for ResearchPlan {
    fn check_contract(&self) -> Result<(), String> {
        if self.section_ids.is_empty() {
            return Err("section_ids must not be empty".into());
        }
        Ok(())
    }
}

impl ContractValid for ResearchSection {
    fn check_contract(&self) -> Result<(), String> {
        if self.title.is_empty() {
            return Err("title must not be empty".into());
        }
        if self.order < 0 {
            return Err(format!("order must be >= 0, got {}", self.order));
        }
        Ok(())
    }
}

impl ContractValid for ResearchTask {
    fn check_contract(&self) -> Result<(), String> {
        if self.idempotency_key.is_empty() {
            return Err("idempotency_key must not be empty".into());
        }
        if self.retry_count.is_some_and(|r| r < 0) {
            return Err("retry_count must be >= 0".into());
        }
        if self.max_retries.is_some_and(|r| r < 0) {
            return Err("max_retries must be >= 0".into());
        }
        Ok(())
    }
}

impl ContractValid for TaskDependency {}
impl ContractValid for ResearchRun {}
